extractor/extractor.go

```diff
@@ -0,0 +1,266 @@
+					ApiVersion: parts[1],
+				})
+			}
+		}
+	}
+
+	return modules, nil
+}
+
+func Services(moduleDirectory string, module *model.Module) ([]model.Service, error) {
+	serviceDirs, err := os.ReadDir(moduleDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("error reading module directory: %w", err)
+	}
+
+	services := []model.Service{}
+
+	for i, path := range serviceDirs {
+		if path.IsDir() {
+			serviceName := path.Name()
+			services = append(services, model.Service{Name: serviceName})
+
+			// Construct the .proto file path
+			protoFilename := fmt.Sprintf("%s_%s%s%s", module.Name, module.ApiVersion[:1], module.ApiVersion[1:], serviceName)
+			protoFilename = cases.Snake(protoFilename)
+			protoFilename = fmt.Sprintf("%s.proto", protoFilename)
+			protoFilePath := filepath.Join(moduleDirectory, serviceName, "proto", protoFilename)
+
+			services[i].Enums, err = getEnums(protoFilePath)
+			if err != nil {
+				return nil, fmt.Errorf("error getting enums: %w", err)
+			}
+
+			modelsFolder := filepath.Join(moduleDirectory, path.Name(), "models", "models.go")
+			services[i].Entities, err = getEntities(modelsFolder)
+			if err != nil {
+				return nil, fmt.Errorf("error getting entities: %w", err)
+			}
+		}
+	}
+
+	return services, nil
+}
+
+func getEnums(protoFilePath string) ([]model.Enum, error) {
+	serviceEnums := []model.Enum{}
+
+	file, err := os.Open(protoFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("error opening .proto file: %w", err)
+	}
+	defer file.Close()
+
+	reader := bufio.NewReader(file)
+
+	proto, _ := protoparser.Parse(reader)
+
+	enumInfo := getEnumInformation(proto)
+
+	for enumName, enumValues := range enumInfo {
+		serviceEnums = append(serviceEnums, model.Enum{Name: enumName, Values: enumValues})
+	}
+	return serviceEnums, nil
+}
+
+func getEnumInformation(proto *parserStructs.Proto) map[string][]string {
+	enumInfo := make(map[string][]string)
+
+	// Traverse ProtoBody and look for elements of type *Enum
+	for _, element := range proto.ProtoBody {
+		if enum, ok := element.(*parserStructs.Enum); ok {
+			enumName := enum.EnumName
+			enumValues := make([]string, len(enum.EnumBody))
+
+			// Extract enum values from EnumBody
+			for i, enumElement := range enum.EnumBody {
+				if enumField, ok := enumElement.(*parserStructs.EnumField); ok {
+					enumValues[i] = enumField.Ident
+				}
+			}
+
+			enumInfo[enumName] = enumValues
+		}
+	}
+
+	return enumInfo
+}
+
+func getEntities(modelsGo string) ([]model.Entity, error) {
+	entities := []model.Entity{}
+
+	fset := token.NewFileSet()
+
+	file, err := parser.ParseFile(fset, modelsGo, nil, parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing models.go: %w", err)
+	}
+
+	// Default to in memory
+	persistence := "memory"
+
+	for _, decl := range file.Decls {
+		if genDecl, ok := decl.(*ast.GenDecl); ok {
+			parseTypeSpecs(genDecl.Specs, &entities, persistence)
+
+			persistence = getPersistence(genDecl.Specs)
```

Comment:
Should getPersistence happen before parseTypeSpecs?
In the current version it looks to me like the persistence we determine for a declaration is only applied to the Entity model of the next declaration, instead of the current one.

Reply:
agree, I'm going to make an overall comment that we need testing for this. We should include tests to cover what @maartenh is talking about ie persistence is taken from previous loop's persistence. I would further recommend that you declare the persistence var inside the loop so its not possible to take the previous state - there doesn't seem to be a reason for it to be scoped where it it...